src/igr-tests/igr_functions

```diff
@@ -0,0 +1,239 @@
+}
+# NOTE: Always quote your error/warning() messages.
+error() {
+    >&2 echo "${TEST_NAME:-}: Error: $1"
+    # Built-in sub_error
+    if [ -n "${2:-}" ]; then
+        >&2 echo " ... $2"
+    fi
+    if [ -n "${STAGE:-}" ]; then
+        echo "${TEST_NAME:-}: Failed at stage $STAGE."
+    fi
+    if [ -e "${SOCKET:-}" ]; then
+        stop_dinit # A dinit instance is running, stopping...
+    fi
+    exit 1
+}
+warning() {
+    echo
+    >&2 echo "${TEST_NAME:-}: Warning: $1"
+    # Built-in sub_warning
+    if [ -n "${2:-}" ]; then
+        >&2 echo " ... $2"
+    fi
+    echo
+}
+
+## Executable path resolvers functions
+# These functions return 0 and set suitable variable when program found.
+# And return 1 on failure.
+find_dinit() {
+    if [ -z "${DINIT:-}" ]; then
+        if [ -x "$EXEC_PREFIX/dinit" ]; then
+            export DINIT="$EXEC_PREFIX/dinit"
+        else
+            return 1
+        fi
+    fi
+    return 0
+}
+find_dinitctl() {
+    if [ -z "${DINITCTL:-}" ]; then
+        if [ -x "$EXEC_PREFIX/dinitctl" ]; then
+            export DINITCTL="$EXEC_PREFIX/dinitctl"
+        else
+            return 1
+        fi
+    fi
+    return 0
+}
+find_dinitcheck() {
+    if [ -z "${DINITCHECK:-}" ]; then
+        if [ -x "$EXEC_PREFIX/dinitcheck" ]; then
+            export DINITCHECK="$EXEC_PREFIX/dinitcheck"
+        else
+            return 1
+        fi
+    fi
+    return 0
+}
+find_dinitmonitor() {
+    if [ -z "${DINITMONITOR:-}" ]; then
+        if [ -x "$EXEC_PREFIX/dinit-monitor" ]; then
+            export DINITMONITOR="$EXEC_PREFIX/dinit-monitor"
+        else
+            return 1
+        fi
+    fi
+    return 0
+}
+
+## Change directory to run-test.sh location
+cd "$(dirname "$0")" || error "Can't change directory to script location."
+
+## Prepare $IGR_OUTPUT
+TEST_NAME="${PWD##*/}"
+if [ -z "$TEST_NAME" ]; then error "Failed to guess test name."; fi
+if [ -z "${IGR_OUTPUT:-}" ]; then
+    IGR_OUTPUT="$PWD"
+    if [ -z "$IGR_OUTPUT" ]; then error "Failed to guess igr output files location."; fi
+    export IGR_OUTPUT
+else
+    # Igr tests are used by meson.
+    mkdir -p "$IGR_OUTPUT/$TEST_NAME/"
+    IGR_OUTPUT="$IGR_OUTPUT/$TEST_NAME/"
+    export IGR_OUTPUT
+fi
+export SOCKET="$IGR_OUTPUT/socket"
+mkdir -p "$IGR_OUTPUT"/output/
+if [ -n "${DEBUG:-}" ]; then
+    QUIET=""
+else
+    QUIET="--quiet"
+fi
+
+## Integration tests helper functions
+# This function spawn a dinit daemon with "$SOCKET" socket path.
+# It's pass its flags as dinit daemon flags.
+# RESULT: return 0 & sets "$DINITPID" variable on success.
+#         throw an error on failure.
+spawn_dinit() {
+    find_dinit || error "Cannot find dinit exec path." "Check 'EXEC_PREFIX' variable value"
+    "$DINIT" $QUIET -u -d sd -p "$SOCKET" "$@" &
+    DINITPID=$!
+    # Wait for Dinit socket shows up.
+    TIMEOUT=0
+    while [ ! -e "$SOCKET" ]; do
+        if [ $TIMEOUT -le 600 ]; then
+            sleep 0.1
+            TIMEOUT=$((TIMEOUT+=1))
```

Comment:
`+=` isn't required to be supported by POSIX (and I think you meant to use just `+` anyway).

Reply:
Actually looks like I'm wrong, POSIX does require `+=`. But since you have assignment you don't need `+=` here.

Reply:
Done